layer1/executor/task_executor.go

```diff
@@ -0,0 +1,354 @@
+		return false, nil
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return true, ctx.Err()
+		case <-time.After(constants.TaskExecutorPoolingTime):
+		}
+
+		if receiptResponse.IsReady() {
+			receipt, err = receiptResponse.GetReceiptBlocking(ctx)
+			if err != nil {
+				var txnStaleError *transaction.ErrTransactionStale
+				if errors.As(err, &txnStaleError) {
+					logger.Info("got a stale transaction and couldn't retry, finishing execution")
+					return false, err
+				}
+				logger.Warnf("got a error while waiting for receipt %v, retrying execution", err)
+				return false, nil
+			}
+
+			if receipt.Status == types.ReceiptStatusSuccessful {
+				logger.Debug("got a successful receipt")
+				return true, nil
+			} else {
+				logger.Warn("got a reverted receipt, retrying")
+				return false, nil
+			}
+		}
+
+		logger.Trace("receipt is not ready yet")
+		hasToExecute, err := shouldExecute(ctx, task)
+		if err != nil {
+			return false, err
+		}
+		if !hasToExecute {
+			return true, nil
+		}
+	}
+}
+
+// shouldExecute checks if a task should be executed. In case of recoverable errors the
+// function is going to retry `constants.MonitorRetryCount` times. If it
+// exhausted the number of retries, it sends true since there's no information.
+// The function returns false in case of unrecoverable errors, or if it
+// shouldn't execute a task. In case of no errors, the return value is true
+// (default case to t.ShouldExecute to return that a task should be executed).
+func shouldExecute(ctx context.Context, task tasks.Task) (bool, error) {
+	logger := task.GetLogger()
+	for i := uint64(1); i <= constants.MonitorRetryCount; i++ {
+		select {
+		case <-ctx.Done():
+			return false, ctx.Err()
+		default:
+		}
+		if hasToExecute, err := task.ShouldExecute(ctx); err != nil {
+			if err.IsRecoverable() {
+				logger.Tracef("got a recoverable error during task.ShouldExecute: %v", err)
+				if err := sleepWithContext(ctx, constants.MonitorRetryDelay); err != nil {
+					return false, err
+				}
+				continue
+			}
+			logger.Debugf("got a non recoverable error during task.ShouldExecute: %v", err)
+			return false, err
+		} else {
+			logger.Tracef("should execute BaseTask: %v", hasToExecute)
+			return hasToExecute, nil
+		}
+	}
+
+	return true, nil
+}
+
+// persistState TaskExecutor to database.
+func (te *TaskExecutor) persistState() error {
+	logger := logging.GetLogger("staterecover").WithField("State", "task_executor")
+	rawData, err := json.Marshal(te)
+	if err != nil {
+		return err
+	}
+
+	err = te.database.Update(func(txn *badger.Txn) error {
+		key := dbprefix.PrefixTaskExecutorState()
+		logger.WithField("Key", string(key)).Debug("Saving state in the database")
+		if err := utils.SetValue(txn, key, rawData); err != nil {
+			logger.Error("Failed to set Value")
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	if err := te.database.Sync(); err != nil {
+		logger.Error("Failed to set sync")
+		return err
```

Comment:
Difficult to test. Ignore coverage report but consider what exit path does.

Reply:
If I got an error during the state persistence I just log it and continue the execution. Do you think I should consider closing the executor if it's out of sync?